### ISSUE TITLE
Create Application Exception/Error Presentation

### ISSUE THREAD

**Issue by petersgiles** [Particular/ServiceControlInstaller#22]
@petersgiles :pig: @gbiellem :chicken: @distantcam :chicken: 

The Installer need a global error catcher and presenter to tell the user when an action failed. A catchall like a 500 error on a web site.

### Plan of Attack
- [x] Create Handler
- [x] Ensure Exceptions are logged to System Event Log
- [x] Pretty Exception Dialog

- [x] report via Raygun with the user's consent (add a button to the exception dialog)

**Comment by petersgiles** [comment]
@distantcam @HEskandari @gbiellem 

I've been thinking about the 'fancy' exception dialog. I'm not sure we can do it. most of the exceptions I get have been with WPF stuff so If the dialog was rendered in WPF and we had a WPF fail wouldn't we end up in a WPF fail loop?

In any case I think we should sideline the fancy dialog idea for now and make it its own nice to have issue for later.

Thoughts?

**Comment by HEskandari** [comment]
The only situation I have seen that exception loop is that if the exception dialog fails to render, e.g. you use a static resource (dynamic resource is more problematic) that doesn't exist and get an exception at runtime, where you try to show the other exception. This should be very rare though.

Maybe @distantcam thinks differently?

**Comment by petersgiles** [comment]
@gbiellem can we call this done now?

**Comment by distantcam** [comment]
@petersgiles No?

So long as the exception dialog is kept isolated it can render just fine while the rest of the app is going bang. The SI error dialog works just fine. So I'm thinking we could just copy that over and rebadge it.

https://github.com/Particular/ServiceInsight/blob/master/src/ServiceInsight/Framework/ExceptionHandler.xaml
https://github.com/Particular/ServiceInsight/blob/master/src/ServiceInsight/Framework/ExceptionHandler.xaml.cs

**Comment by petersgiles** [comment]
Ok will put this in

**Comment by petersgiles** [comment]
@gbiellem @distantcam this is in place now. just waiting on Dispatcher issue to be resolved.

**Comment by petersgiles** [comment]
![exception](https://cloud.githubusercontent.com/assets/1644481/9319696/8262d192-4594-11e5-9640-99ae6b78753f.PNG)

@sergioc here is the current exception screen...do you want to add any design flair

**Comment by distantcam** [comment]
I've fixed the dispatcher problem, so that should be good.

Do we want to add a report button just to make it easier to report errors when the server is internet connected?

**Comment by petersgiles** [comment]
@distantcam it is hooked up to raygun so it does it automatically

**Comment by sergioc** [comment]
@petersgiles I'd like to improve the presentation, but I have some questions:

- Are we always able to show a precise error message like the one at the top in the red container?
  - If yes, or for those we know, is there a list of messages?
- Are there more exception exception navigation groups at the root level besides "Exception"?
- Can you copy-paste here an example stack trace thrown by SC?
- Are you sure it's a good idea to automatically report (via Raygun), i.e. without the user's consent?

**Comment by distantcam** [comment]
> Are we always able to show a precise error message like the one at the top in the red container?
Yes, but it's an internal error anyway, so I'm not sure how much use it will be to the user.

> Are there more exception navigation groups at the root level besides "Exception"?
Possibly. Again this should just be internal data for us.

> Are you sure it's a good idea to automatically report (via Raygun), i.e. without the user's consent?
I don't think it's a good idea. They may be running this from a production machine and may not want details about that machine sent out.

**Comment by gbiellem** [comment]
+1 for asking for permission to send to Raygun.

**Comment by sergioc** [comment]
>> Are we always able to show a precise error message like the one at the top in the red container?

>Yes, but it's an internal error anyway, so I'm not sure how much use it will be to the user.

>> Are there more exception navigation groups at the root level besides "Exception"?

> Possibly. Again this should just be internal data for us.

If all of this is internal info for us and of no help for the user, then I wouldn't show any detailed info. It's best to only show an error message, with a title as precise as possible (i.e. the msg in the red container) in order not to leave the user clueless, and make it easy to report the error + contact support.

**Comment by petersgiles** [comment]
> Are we always able to show a precise error message like the one at the top in the red container?
> If yes, or for those we know, is there a list of messages?

It's an Unhandled exception dialog, so not really. However we do throw a bunch of them as part of checking for complex validations. We control the 'Message' in those.

>    Are there more exception navigation groups at the root level besides "Exception"?

The Exception root is a product of the treeview control needing a root. There's probably a way to improve it. Reading other comments we could get rid of it altogether. As a techie I like stacktraces but maybe an ops person wouldn't really care for it.

> Are you sure it's a good idea to automatically report (via Raygun), i.e. without the user's consent?

Yep good point...we can add a button to do it.

**Comment by sergioc** [comment]
So what's the bare minimum message we could display in the event of an unhandled exception occurring?

**Comment by sergioc** [comment]
I think the showing the exception can still be useful, but not by default. It would be useful in order to allow the user to see what info would we collect, should they opt-in for reporting.

**Comment by petersgiles** [comment]
@sergioc 

We've added a screen that displays the messages we 'know' about including deeper validations and other things 'like service not starting' we call it the 'action report card' It's a little ugly right now but @distantcam is helping us make it look consistent with other dialogs.

The Unhandled Exceptions, that is, errors that we were unable to predict', come to this screen where we feel that showing the details of the exception is really important. The rule of thumb should be that users of this product should rarely ever see this screen as we mange all other edge events through other mechanisms. For example the "Nested Path are not supported..." message in the example exception diagram we posted earlier would never appear in this screen now.

> So what's the bare minimum message we could display in the event of an unhandled exception occurring?

I guess we could show just the text in red but if we also wanted to 'reveal' more detail that would involve buttons and other UI elements that we would have to maintain. I believe, as someone who uses products like these a lot, that users would ALWAYS click to reveal more detail. So why not just show them in the first place and save the trouble of maintaining more code?

**Comment by gbiellem** [comment]
@distantcam  can you provide a screen shot of the report card once it's cleaned up so @sergioc  has visibility of what we've done.

And I'm :+1:  for keeping the details visible in the unhandled exception dialog.  Primarily so the user sees what is being sent back to us if they report via raygun.

**Comment by distantcam** [comment]
Here's the report card.

![image](https://cloud.githubusercontent.com/assets/1298847/9375347/1fcdd186-4732-11e5-9cfb-fefccbc31da3.png)

**Comment by petersgiles** [comment]
@distantcam this is excellent. I think the functionality we want is here.

@sergioc I'm closing this issue since we have the functionality we want now. Please feel free to raise a new issue if you want to address a specific issue.
